sdks/go/pkg/beam/io/synthetic/source.go

```diff
@@ -0,0 +1,151 @@
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package synthetic contains transforms for creating synthetic pipelines.
+// Synthetic pipelines are pipelines that simulate the behavior of possible
+// pipelines in order to test performance, splitting, liquid sharding, and
+// various other infrastructure used for running pipelines. This category of
+// tests are not concerned with the correctness of the elements themselves, but
+// need to simulate transforms that output many elements throughout varying
+// pipeline shapes.
+package synthetic
+
+import (
+	"github.com/apache/beam/sdks/go/pkg/beam"
+	"github.com/apache/beam/sdks/go/pkg/beam/io/rtrackers/offsetrange"
+	"math/rand"
+	"time"
+)
+
+// Source creates a synthetic source transform that emits randomly
+// generated KV<[]byte, []byte> elements.
+//
+// This transform accepts a PCollection of SourceConfig, where each SourceConfig
+// determines the synthetic source's behavior for that element.
+//
+// This transform outputs a PCollection of randomly generated
+// KV<byte[], byte[]> elements.
+func Source(s beam.Scope, col beam.PCollection) beam.PCollection {
+	s = s.Scope("synthetic.Source")
+
+	return beam.ParDo(s, sourceFn{}, col)
+}
+
+// syntheticSourceFn is a splittable DoFn implementing behavior for synthetic
+// sources. For usage information, see `SyntheticSource`.
+//
+// The syntheticSourceFn is expected to receive elements of type sourceConfig
+// and follow that config to determine its behavior when splitting and emitting
+// elements.
+type sourceFn struct {
+	rng *rand.Rand
+}
+
+// CreateInitialRestriction creates an offset range restriction representing
+// the number of elements to emit.
+func (fn *sourceFn) CreateInitialRestriction(config SourceConfig) offsetrange.Restriction {
+	return offsetrange.Restriction{
+		Start: 0,
+		End:   int64(config.NumElements),
+	}
+}
+
+// SplitRestriction splits restrictions equally according to the number of
+// initial splits specified in SourceConfig. Each restriction output by this
+// method will contain at least one element, so the number of splits will not
+// exceed the number of elements.
+func (fn *sourceFn) SplitRestriction(config SourceConfig, rest offsetrange.Restriction) (splits []offsetrange.Restriction) {
+	if config.InitialSplits <= 1 {
+		// Don't split, just return original restriction.
+		splits = append(splits, rest)
+		return splits
+	}
+
+	num := int64(config.InitialSplits)
```

Comment:
Good point, I should've thought of that. I don't really wanna do that in this PR though because I want to write some tests for the method as well, which has just made me realize that the offsetrange package as a whole probably needs tests. So I'll just leave a TODO for that and make a new PR for it.

Reply:
+1 for a TODO for this PR. Thanks!